File: Lua/shell_sort.lua
function shell_sort(arr)
    local n = #arr
    local gap = math.floor(n / 2)

    while gap > 0 do
        for i = gap + 1, n do
            local temp = arr[i]
            local j = i
            while j > gap and arr[j - gap] > temp do
                arr[j] = arr[j - gap]
                j = j - gap
            end
            arr[j] = temp
        end
        gap = math.floor(gap / 2)
    end
end

local arr = {5, 1, 4, 2, 8}
shell_sort(arr)
print(table.concat(arr, " "))

File: Java/ShellSort.java
public class ShellSort {
    public static void shellSort(int[] arr) {
        int n = arr.length;
        for (int gap = n / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
            int temp = arr[i];
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap) {
                arr[j] = arr[j - gap];
            }
            arr[j] = temp;
            }
        }
        }

        public static void main(String[] args) {
        int[] arr = {12, 34, 54, 2, 3};

        shellSort(arr);

        System.out.println("Array after sorting:");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        }
    }
